src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs

```diff
@@ -0,0 +1,269 @@
+        public static bool IsLambdaBodyStatementOrExpression(SyntaxNode node)
+        {
+            return IsLambdaBody(node);
+        }
+
+        public static bool IsLambdaBodyStatementOrExpression(SyntaxNode node, out SyntaxNode lambdaBody)
+        {
+            lambdaBody = node;
+            return IsLambdaBody(node);
+        }
+
+        /// <summary>
+        /// If the specified node represents a lambda returns a node (or nodes) that represent its body (bodies).
+        /// </summary>
+        public static bool TryGetLambdaBodies(SyntaxNode node, out SyntaxNode lambdaBody1, out SyntaxNode lambdaBody2)
+        {
+            lambdaBody1 = null;
+            lambdaBody2 = null;
+
+            switch (node.Kind())
+            {
+                case SyntaxKind.ParenthesizedLambdaExpression:
+                case SyntaxKind.SimpleLambdaExpression:
+                case SyntaxKind.AnonymousMethodExpression:
+                    lambdaBody1 = ((AnonymousFunctionExpressionSyntax)node).Body;
+                    return true;
+
+                case SyntaxKind.FromClause:
+                    // The first from clause of a query expression is not a lambda.
+                    if (node.Parent.IsKind(SyntaxKind.QueryExpression))
+                    {
+                        return false;
+                    }
+
+                    lambdaBody1 = ((FromClauseSyntax)node).Expression;
+                    return true;
+
+                case SyntaxKind.JoinClause:
+                    var joinClause = (JoinClauseSyntax)node;
+                    lambdaBody1 = joinClause.LeftExpression;
+                    lambdaBody2 = joinClause.RightExpression;
+                    return true;
+
+                case SyntaxKind.LetClause:
+                    lambdaBody1 = ((LetClauseSyntax)node).Expression;
+                    return true;
+
+                case SyntaxKind.WhereClause:
+                    lambdaBody1 = ((WhereClauseSyntax)node).Condition;
+                    return true;
+
+                case SyntaxKind.AscendingOrdering:
+                case SyntaxKind.DescendingOrdering:
+                    lambdaBody1 = ((OrderingSyntax)node).Expression;
+                    return true;
+
+                case SyntaxKind.SelectClause:
+                    lambdaBody1 = ((SelectClauseSyntax)node).Expression;
+                    return true;
+
+                case SyntaxKind.GroupClause:
+                    var groupClause = (GroupClauseSyntax)node;
+                    lambdaBody1 = groupClause.GroupExpression;
+                    lambdaBody2 = groupClause.ByExpression;
+                    return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// "Pair lambda" is a synthesized lambda that creates an instance of an anonymous type representing a pair of values. 
+        /// TODO: Avoid generating these lambdas. Instead generate a method on the anonymous type, or use KeyValuePair instead.
```

Comment:
Please do not put TODO in the code. Add an issue on github instead. We're going to have to go through the sources and remove TODOs soon. It is better to add the github issue as soon as the work item is understood.

Also, I have no idea what this TODO is trying to say.